subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/VersionedNativeCompiler.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.nativeplatform.toolchain.internal;
+
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.language.base.internal.compile.Compiler;
+import org.gradle.nativeplatform.toolchain.NativeCompilerVersion;
+import org.gradle.util.VersionNumber;
+
+public class VersionedNativeCompiler<T extends NativeCompileSpec> implements Compiler<T> {
```

Comment:
Nit-pick: `VersionAwareNativeCompiler` might be a better name perhaps.

Reply:
More importantly, it should either implement `NativeCompiler`, which should be extracted as an interface. Alternatively it could be called `VersionAwareCompiler` and implement `Compiler`, but then it should be in `org.gradle.language.base.internal.compile`. I think I like the latter option better: there's nothing native-specific about this class.

The latter option would also mean we'd need to have a general `CompilerVersion` type instead of `NativeCompilerVersion`. I think I like that, too. Either way, we should make the types here consistent: either all native, or no native at all.

Reply:
I renamed the type to `VersionAwareCompiler` and moved it to `org.gradle.language.base.internal.compile`. I put `CompilerVersion` into `org.gradle.language.base.compile`. WDYT?

Reply:
:+1: